Batch container block settings action calls

youknowriad opened the pull request:
## What?

It seems that the introduction of the list block with inner blocks highlighted an issue with the initial rendering of container blocks. (You can see that the loading metric increased by 200%). 

The issue is that each time a container block is rendered a `updateBlockListSettings` action is triggered calling a potential attempt to re-render the components on the page. So if you load a post with 50 container blocks, it will try to re-rendered the components on the page 50 times.

## How?

This PR attempts to solve that by batching the calls to `updateBlockListSettings` to ensure they only trigger a single re-render.

## Testing Instructions

The performance job should tell us whether this PR solves the issue or not.

youknowriad committed:
Batch container block settings action calls

youknowriad commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1435,7 +1435,7 @@ export function getTemplateLock( state, rootClientId ) {
 	const blockListSettings = getBlockListSettings( state, rootClientId );
 	if ( ! blockListSettings ) {
-		return null;
+		return undefined;
```
Here I'm making the assumption that `null` and `undefined` are the same thing. Maybe @jorgefilipecosta can confirm. 

The `updateBlockListSettings` was being called twice for each container block because of this.

jorgefilipecosta replied:
Yes, it should be the same thing.

youknowriad commented:
Here's the result of the PR so far. The numbers improve a lot but we're not back entirely to the previous results (before list block). 

<img width="634" alt="Screen Shot 2022-09-08 at 7 12 29 AM" src="https://user-images.githubusercontent.com/272444/189047681-3dfd6c62-1def-4765-96f8-7a6e30d7e875.png">

It's important to note that now the test post contains 1400 blocks where previously it was 1000. The reason is because when the post is loaded the content is migrated and list blocks with inner blocks are created. That difference in the number of blocks mean there are more components to render causing an impact as well on the numbers.

It also looks like there are some e2e test failures due to my optimizations, which I'll have to figure out before merging this.

youknowriad committed:
Fix unit tests

youknowriad committed:
Fix focus loss on inserting quote blocks

youknowriad committed:
Fix ctrl + click to insert quotes

youknowriad committed:
Revert the initial stop block selection bug

ellatrix commented on packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:
```diff
@@ -98,7 +101,29 @@ export default function useNestedSettingsUpdate(
 		}
 
 		if ( ! isShallowEqual( blockListSettings, newSettings ) ) {
```
What setting causes `isShallowEqual` to be false?

youknowriad replied:
none, it's just the first time this gets called, it needs to be batched across the 500 blocks that call it in the same render commit.

ellatrix replied:
Oh, ok. Why is that not initialised/preloaded in the store though?

ellatrix replied:
Sorry, am bit slow today. Because the information is in a React component of course.

ellatrix commented on packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js:
```diff
@@ -68,19 +68,26 @@ export default function useInnerBlockTemplateSync(
 					template
 				);
+
 				if ( ! isEqual( nextBlocks, innerBlocks ) ) {
```
How come `isEqual` returns false so much here? Blocks shouldn't change during load?

youknowriad replied:
It doesn't, the change here was a consequence of the change above to fix `ctrl + click` in the inserter.

youknowriad committed:
Fix zombie bug

youknowriad commented:
Hey @geriux This PR improves the performance of container blocks drastically but it seems I have a failing mobile unit test. Would you have time to help me understand and fix this failure? Thank you.

youknowriad committed:
Fix list e2e test

geriux commented:
> Hey @geriux This PR improves the performance of container blocks drastically but it seems I have a failing mobile unit test. Would you have time to help me understand and fix this failure? Thank you.

Hey there 👋 I'll check it out 👍

youknowriad committed:
Fix block migrations on load

geriux commented:
@youknowriad I have a fix for the tests that are failing, I also tested and the list block works correctly with these new changes 🎉 . Do you want me to push those to your branch or should I create a PR using this branch as base? Let me know 😃 

Thanks!

youknowriad commented:
Go push here. Thanks @geriux

Commit:
Mobile - List block - Updates tests:

- Remove V1 List block tests
- Updates the insertion test to add triggering of the onLayout event for the block list

geriux commented:
Mobile unit tests are passing now 🚀 let us know if you have any other issues. Thanks!

youknowriad committed:
Fix more e2e tests

youknowriad commented:
@geriux Looks like there's a new failure (group block) this time. 😬

geriux commented:
> @geriux Looks like there's a new failure (group block) this time. 😬

This should fix it https://github.com/WordPress/gutenberg/pull/44261 but I'm looking for a reviewer :D

geriux commented:
> This should fix it #44261 but I'm looking for a reviewer :D

It is now merged ✅

mtias commented on packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:
```diff
@@ -98,7 +101,29 @@ export default function useNestedSettingsUpdate(
 
 		if ( ! isShallowEqual( blockListSettings, newSettings ) ) {
-			updateBlockListSettings( clientId, newSettings );
+			// This is an optimization for the initial rendering of a block list
```
This comment reads a bit oddly to me. A suggestion:

```suggestion
			// Batch updates to block list settings to avoid triggering cascading renders 
			// for each container block included in a tree and optimize initial render.
```

And remove the rest

youknowriad committed:
Better comment

Co-authored-by: Matias Ventura <mv@matiasventura.com>

ntsekouras commented on packages/block-editor/src/components/inner-blocks/use-inner-block-template-sync.js:
```diff
@@ -47,42 +47,53 @@ export default function useInnerBlockTemplateSync(
-		) {
+		// There's an implicit dependency between useInnerBlockTemplateSync and useNestedSettingsUpdate
+		// The former needs to happen after the latter and since the latter is using microtasks to batch updates (performance optimization),
+		// We need to schedule this one in a microtask as well.
```
```suggestion
		// we need to schedule this one in a microtask as well.
```

ntsekouras commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1714,9 +1714,12 @@ export function automaticChangeStatus( state, action ) {
 			return state;
 	}
 
+	// TODO: This is a source of bug, as each time there's a change in timing,
```
I think we need to add more info here about what the bug actually is, for future reference. Maybe the link to the PR? 🤔

youknowriad replied:
There's no identified bug, it's just a bad pattern (returning undefined from a reducer) that can cause bugs quickly and these bugs are very hard to track.

ntsekouras reviewed (approved):
Thanks Riad!

youknowriad committed:
copy

Co-authored-by: Nik Tsekouras <ntsekouras@outlook.com>

ockham commented:
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: c21b6263c0
